OpenRA.Mods.Common/Warheads/FireClusterWarhead.cs

```diff
@@ -0,0 +1,126 @@
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.GameRules;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Warheads
+{
+	public class FireClusterWarhead : Warhead, IRulesetLoaded<WeaponInfo>
+	{
+		[WeaponReference, FieldLoader.Require]
+		[Desc("Has to be defined in weapons.yaml as well.")]
+		public readonly string Weapon = null;
+
+		[Desc("Number of weapons fired at random 'x' cells. Negative values will result in a number equal to 'x' footprint cells fired.")]
+		public readonly int RandomClusterCount = -1;
+
+		[FieldLoader.Require]
+		[Desc("Size of the cluster footprint")]
+		public readonly CVec Dimensions = CVec.Zero;
+
+		[FieldLoader.Require]
+		[Desc("Cluster footprint. Cells marked as X will be attacked.",
+			"Cells marked as x will be attacked randomly until RandomClusterCount is reached.")]
+		public readonly string Footprint = string.Empty;
+
+		WeaponInfo weapon;
+
+		public void RulesetLoaded(Ruleset rules, WeaponInfo info)
+		{
+			if (!rules.Weapons.TryGetValue(Weapon.ToLowerInvariant(), out weapon))
+				throw new YamlException("Weapons Ruleset does not contain an entry '{0}'".F(Weapon.ToLowerInvariant()));
+		}
+
+		public override void DoImpact(Target target, Actor firedBy, IEnumerable<int> damageModifiers)
+		{
+			if (!target.IsValidFor(firedBy))
+				return;
+
+			var map = firedBy.World.Map;
+			var targetCell = map.CellContaining(target.CenterPosition);
+
+			var targetCells = CellsMatching(targetCell, false);
+			foreach (var c in targetCells)
+				FireProjectileAtCell(map, firedBy, target, c);
+
+			if (RandomClusterCount != 0)
+			{
+				var randomTargetCells = CellsMatching(targetCell, true);
+				var clusterCount = RandomClusterCount < 0 ? randomTargetCells.Count() : RandomClusterCount;
+				if (randomTargetCells.Any())
+					for (var i = 0; i < clusterCount; i++)
+						FireProjectileAtCell(map, firedBy, target, randomTargetCells.Random(firedBy.World.SharedRandom));
+			}
+		}
+
+		void FireProjectileAtCell(Map map, Actor firedBy, Target target, CPos targetCell)
+		{
+			var tc = Target.FromCell(firedBy.World, targetCell);
+
+			if (!weapon.IsValidAgainst(tc, firedBy.World, firedBy))
+				return;
+
+			var args = new ProjectileArgs
+			{
+				Weapon = weapon,
+				Facing = (map.CenterOfCell(targetCell) - target.CenterPosition).Yaw.Facing,
+
+				DamageModifiers = !firedBy.IsDead ? firedBy.TraitsImplementing<IFirepowerModifier>()
+					.Select(a => a.GetFirepowerModifier()).ToArray() : new int[0],
+
+				InaccuracyModifiers = !firedBy.IsDead ? firedBy.TraitsImplementing<IInaccuracyModifier>()
```

Comment:
IMO this could be ignored to complete the set - why should the current state of the actor which may have fired this many ticks ago have any impact on the spread properties?